kpi-tracker/src/main.rs

```diff
@@ -134,69 +187,143 @@ async fn accounts_in_block(
+                    events.push(event);
+                }
+                _ => {}
+            };
+        }
+        AccountCreation(act) => {
+            let address = act.address;
+            let details = account_details(block_hash, act);
+
+            println!("ACCOUNT:\naddress: {}\ndetails: {:?}", &address, &details); // Logger debug
+
+            let block_event = BlockEvent::AccountCreation(address, details);
+            events.push(block_event);
+        }
+        _ => {}
+    };
+
+    events
+}
+
+/// Maps a stream of transactions to a stream of `BlockEvent`s
+async fn get_block_events(
     node: &mut Client,
     block_hash: BlockHash,
-) -> anyhow::Result<BTreeMap<AccountAddress, AccountDetails>> {
-    let mut transactions = node
+) -> anyhow::Result<impl Stream<Item = anyhow::Result<BlockEvent>>> {
+    let transactions = node
         .get_block_transaction_events(block_hash)
         .await
-        .context(format!(
-            "Could not get transactions for block: {}",
-            block_hash
-        ))?
+        .with_context(|| format!("Could not get transactions for block: {}", block_hash))?
         .response;
 
-    let mut new_accounts = BTreeMap::new();
-
-    while let Some(res) = transactions.next().await {
-        let transaction = res.with_context(|| {
-            format!(
-                "Error while streaming transactions for block {}",
-                block_hash
+    let block_events = transactions
+        .map_ok(move |bi| {
+            let block_events: Vec<Result<BlockEvent, anyhow::Error>> =
+                to_block_events(block_hash, bi)
+                    .into_iter()
+                    .map(Ok)
+                    .collect();
+            futures::stream::iter(block_events)
+        })
+        .map_err(move |err| {
+            anyhow!(
+                "Error while streaming transactions for block: {} - {}",
+                block_hash,
+                err
             )
-        })?;
-
-        match transaction.details {
-            AccountCreation(act) => {
-                let address = act.address;
-                let account_details = account_details(block_hash, Some(act));
-
-                new_accounts.insert(address, account_details);
-            }
-            _ => {}
-        }
-    }
+        })
+        .try_flatten();
 
-    Ok(new_accounts)
+    Ok(block_events)
 }
 
 /// Insert as a single DB transaction to facilitate easy recovery, as the service can restart from
 /// current height stored in DB.
 fn update_db(
     db: &mut DB,
-    (block_hash, block_details): (BlockHash, &BlockDetails),
+    (block_hash, block_details): (BlockHash, BlockDetails),
     accounts: BTreeMap<AccountAddress, AccountDetails>,
+    transactions: Option<BTreeMap<TransactionHash, TransactionDetails>>,
+    contract_modules: Option<BTreeMap<ModuleRef, ContractModuleDetails>>,
+    contract_instances: Option<BTreeMap<ContractAddress, ContractInstanceDetails>>,
```

Comment:
I'm not sure what the meaning of all these options is.

Reply:
the purpose is to be able to update the database with only accounts (i.e. initialize DB with accounts in genesis block). Maybe it's a bit silly. Could just add them outside of the function in that single case.